common/persistence/sql/sqlTaskPersistence.go

```diff
@@ -51,21 +115,243 @@ func (m *sqlTaskManager) Close() {
 }
 
 func (m *sqlTaskManager) LeaseTaskList(request *persistence.LeaseTaskListRequest) (*persistence.LeaseTaskListResponse, error) {
-	return &persistence.LeaseTaskListResponse{&persistence.TaskListInfo{}}, nil
+	var row tasksListsRow
+	var rangeID int64 = 0
+	var ackLevel int64 = 0
+	if err := m.db.Get(&row, getTaskListSQLQuery, request.DomainID, request.TaskList, request.TaskType); err != nil {
+		if err == sql.ErrNoRows {
+			// The task list does not exist. Create it.
+			if _, err := m.db.NamedExec(createTaskListSQLQuery,
+				tasksListsRow{
+					request.DomainID,
+					rangeID+1,
+					request.TaskList,
+					int64(request.TaskType),
+					ackLevel,
+					int64(request.TaskListKind),
+					MaximumExpiryTs,
+				}); err != nil {
+				return nil, &workflow.InternalServiceError{
+					Message: fmt.Sprintf("LeaseTaskList operation failed. Failed to make task list %v of type %v. Error: %v", request.TaskList, request.TaskType, err),
+				}
+			}
+		}
+	} else {
+		// The task list exists.
+		rangeID = row.RangeID
+		ackLevel = row.AckLevel
+
+		tx, err := m.db.Beginx()
+		if err != nil {
+			return nil, &workflow.InternalServiceError{
+				Message: fmt.Sprintf("LeaseTaskList operation failed. Failed to begin transaction. Error: %v", err),
+			}
+		}
+		defer tx.Rollback()
+
+		if err := lockAndCheckTaskListRangeID(tx, request.DomainID, request.TaskList, request.TaskType, row.RangeID); err != nil {
+			switch err.(type) {
+			case *persistence.ConditionFailedError:
+				return nil, err
+			default:
+				return nil, &workflow.InternalServiceError{
+					Message: fmt.Sprintf("LeaseTaskList operation failed. Error: %v", err),
+				}
+			}
+		}
+
+		if result, err := tx.NamedExec(updateTaskListSQLQuery,
```

Comment:
single update does not need locking and transaction?

Reply:
similar below

Reply:
(We need to separately check the condition and do the update because we want to throw different error codes. Since we need to do things separately, we need a lock.)

Reply:
Please add a comment in the code on this.